File: leetcode/1047_RemoveAllAdjacentDuplicatesInString/lc1047.py
class Solution:
    def removeDuplicates(self, S: str) -> str:
        stack = []
        for ch in S:
            if stack and stack[-1] == ch:
                stack.pop()
            else:
                stack.append(ch)
        return ''.join(stack)
        #top = 0
        #for ch in S:
        #    if top == 0 or S[top-1] != ch:
        #        S = S[:top] + ch + S[top+1:]
        #        top += 1
        #    else:
        #        top -= 1
        #return S[:top]

File: leetcode/1047_RemoveAllAdjacentDuplicatesInString/lc1047.cpp
class Solution {
public:
    string removeDuplicates(string S) {
        int top = 0; // record valid position
        for (char ch : S) {
            if (top == 0 || S[top-1] != ch) {
                S[top++] = ch; // remove different ch to top valid position
            } else {
                top--; // valid, keep extending search area
            }
        }
        S.resize(top);
        return S;
    }
};
